src/stirling/source_connectors/perf_profiler/notes.md

```diff
@@ -0,0 +1,209 @@
+# Stirling profiler notes
+
+### Perf profiler organization
+TODO(jps): describe the various source code files and their respective purposes.
+
+### Symbolization notes
+TODO(jps): cover symbolization for compiled & Java and symbol caching.
+
+### Missing stack frames
+While investigating issue [#719](https://github.com/pixie-io/pixie/issues/719), we noticed that infrequently (less than 0.5% of the time, e.g.) we observed a stack trace with a missing stack frame. Fixing the test case to accommodate the issue was easy, but not informative. Here, we document our investigation into the root cause of the issue.
+
+##### The issue
+The missing stack frames we observed were caused by stack trace samples captured at function entry:
+```
+push   %rbp
+mov    %rsp, %rbp
+```
+or at function exit:
+```
+pop    %rbp
+ret
+```
+If the stack trace is sampled exactly when the program counter points to either the `push` or `ret` instructions shown above, then the stack trace (when walked using frame pointers) will start from the caller's ancestor function: the stack trace will skip the caller of the leaf function and wrongly connect the leaf function to its caller's calling function.
+
+##### Walking back through the stack to find a stack trace
+To understand why, we first needed to understand how the kernel walks back through the stack [1] to reconstruct a stack trace.
+
+Here we show a sketch of the algorithm used by the [kernel stack walker](https://github.com/torvalds/linux/blob/e1c04510f521e853019afeca2a5991a5ef8d6a5b/arch/x86/events/core.c#L2864):
+1. Register `rbp` points to the saved frame base pointer of the calling function. Dereference `rbp` and store in a local variable (we will use variable name `fp` for the saved frame pointer).
+2. Read register `rip`, the instruction pointer. This is the leaf address in the stack trace.
+```
+stack_trace.push_back(rip);
+```
+3. The function return address is always the last thing pushed to the stack before calling a new function. Thus, the function return address is at memory location `rbp + 8` [2]. Read the function return address and add it to the stack trace.
+```
+fn_return_address = *(fp + 8);
+stack_trace.push_back(fn_return_address);
+```
+4. The saved base pointers each, in turn, point to their caller's stack. Find the next caller's base pointer:
+5. If `saved_rbp` is not a valid stack address, exit, else goto step (3).
+
+In pseudo code:
+```
+fp = *rbp;
+stack_trace.push_back(rip);
+while(true) {
+  fn_return_address = *(saved_rbp + 8);
```

Comment:
I realize this is pseudo-code, but shouldn't `saved_rbp` be updated to be equal to `fp` on each loop iteration?

Reply:
Yes, that is true! I had that variable named `saved_rbp` originally, but then thought it would be good to align to Linux source code. Going with: `s/saved_rbp/fp`.